push/src/main/java/org/aerogear/mobile/push/PushService.java

```diff
@@ -0,0 +1,313 @@
+        String hashedCrentials =
+                        Base64.encodeToString(unhashedCredentials.getBytes(), Base64.NO_WRAP);
+        return headerValueBuilder.append(hashedCrentials).toString();
+    }
+
+
+    /**
+     * When a push message is received, all main thread handlers will be notified on the main(UI)
+     * thread. This is very convenient for Activities and Fragments.
+     *
+     * @param handler a handler to added to the list of handlers to be notified.
+     */
+    public static void registerMainThreadHandler(final MessageHandler handler) {
+        nonNull(handler, "handle");
+        MAIN_THREAD_HANDLERS.add(handler);
+    }
+
+    /**
+     * When a push message is received, all background thread handlers will be notified on a non UI
+     * thread. This should be used by classes which need to update internal state or preform some
+     * action which doesn't change the UI.
+     *
+     * @param handler a handler to added to the list of handlers to be notified.
+     */
+    public static void registerBackgroundThreadHandler(final MessageHandler handler) {
+        nonNull(handler, "handle");
+        BACKGROUND_THREAD_HANDLERS.add(handler);
+    }
+
+    /**
+     * This will remove the given handler from the collection of main thread handlers. This MUST be
+     * called when a Fragment or activity is backgrounded via onPause.
+     *
+     * @param handler a handler to be removed to the list of handlers
+     */
+    public static void unregisterMainThreadHandler(final MessageHandler handler) {
+        nonNull(handler, "handle");
+        MAIN_THREAD_HANDLERS.remove(handler);
+    }
+
+    /**
+     * This will remove the given handler from the collection of background thread handlers.
+     *
+     * @param handler a handler to be removed to the list of handlers
+     */
+    public static void unregisterBackgroundThreadHandler(final MessageHandler handler) {
+        nonNull(handler, "handle");
+        BACKGROUND_THREAD_HANDLERS.remove(handler);
+    }
+
+    /**
+     * This will deliver an message to all registered handlers.
+     *
+     * @param message the message to pass
+     */
+    public static void notifyHandlers(final Context context, final Map<String, String> message) {
+        nonNull(context, "context");
+
+        if (BACKGROUND_THREAD_HANDLERS.isEmpty() && MAIN_THREAD_HANDLERS.isEmpty()
+                        && defaultHandler != null) {
+            new AppExecutors().mainThread().execute(() -> defaultHandler.onMessage(context,
+                            Collections.unmodifiableMap(message)));
+        } else {
+
+            for (final MessageHandler handler : BACKGROUND_THREAD_HANDLERS) {
+                new AppExecutors().mainThread().execute(() -> handler.onMessage(context,
+                                Collections.unmodifiableMap(message)));
+            }
+
+            Looper main = Looper.getMainLooper();
```

Comment:
@secondsun Should we also replace `Looper` by `AppExecutors`

Reply:
@danielpassos Yes